synapse/handlers/account_data.py

```diff
@@ -171,7 +171,13 @@ def get_current_key(self, direction: str = "f") -> int:
         return self.store.get_max_account_data_stream_id()
 
     async def get_new_events(
-        self, user: UserID, from_key: int, **kwargs: Any
+        self,
+        user: UserID,
+        from_key: int,
+        limit: Optional[int],
+        room_ids: Collection[str],
+        is_guest: bool,
+        explicit_room_id: Optional[str] = None,
```

Comment:
What are these parameters? They seem to my eye unused. Is it not possible to remove them?

Reply:
Ah, I see, these sources are passed to a for loop in:

https://github.com/matrix-org/synapse/blob/858363d0b7e58fd71875b25d183537bb3b5a397f/synapse/notifier.py#L594-L601

In the `EventSources` class, we seem to keep a list of these but they're typed as `Any`. I think it would be nice to give these a common supertype that specifies the signature of `get_new_event` — I'm happy to take this up in a separate PR though.

Reply:
(actually, I see you already reworked `EventSources` quite a bit)

Reply:
> We keep a list of what?

Event sources

> mypy should know at this point it is one of the sub-types since get_sources returns a Union.

Yes, I should have read on further before commenting since you changed this at the same time :p

> I thought about having a Protocol for this, yes.

This is just my personal preference, but a `Protocol` (i.e. structural subtyping, essentially 'duck typing') seems too weak for this case when we could use 'nominal subtyping' (i.e. actually extend some base class `EventSource`).
I think this would also help IDE completion of 'Find Usages' etc, which is one of the most useful tools (afaik) for a newcomer like me — when it works.

One thing I wonder is whether mypy will complain if you override a method in an incompatible way? I think this would perhaps lead to clearer messages than if you messed up the method signature and it suddenly no longer counts as a subtype of the Protocol.